File: src/_03_BehavioralPattern/_03_08_State/OnlineCourse.js
/**
 * @typedef { import("./Student").default } Student
 * @typedef { import("./State").default } State
 */

export default class OnlineCourse {
  /** @type { State } */
  _state;
  
  /** @type { Student[] } */
  _students;

  /** @type { string[] } */
  _reviews;

  constructor() {
    this._students = [];
    this._reviews = [];
  }
  
  /** @returns { State } */
  getState() {
    return this._state;
  }

  /** @returns { Student[] } */
  getStudents() {
    return this._students;
  }

  /** @returns { string[] } */
  getReviews() {
    return this._reviews;
  }

  /** @param { State } state */
  changeState(state) {
    this._state = state;
  }

  /** @param { Student } student */
  addStudent(student) {
    this._state.addStudent(student);
  }

  /**
   * @param { Student } student
   * @param { string } review
   */
  addReview(student, review) {
    this._state.addReview(student, review);
  }
}

File: src/_03_BehavioralPattern/_03_08_State/java/after/OnlineCourse.java
package _03_BehavioralPattern._03_08_State.java.after;

import java.util.ArrayList;
import java.util.List;

public class OnlineCourse {
  
  private State state;
  
  private List<Student> students = new ArrayList<>();

  private List<String> reviews = new ArrayList<>();

  public void addStudent(Student student) {
    this.state.addStudent(student);
  }

  public void addReview(Student student, String review) {
    this.state.addReview(student, review);
  }

  public State getState() {
    return this.state;
  }

  public List<Student> getStudents() {
    return this.students;
  }

  public List<String> getReviews() {
    return this.reviews;
  }

  public void changeState(State state) {
    this.state = state;
  }
  
}
